[12.0][MIG] delivery_carrier_label_batch

TDu opened the pull request.

yvaucher committed:
[9.0] port module delivery_carrier_label_dispatch

- change name to batch to match with change of `picking_dispatch` name to
`stock_batch_picking`
- mode description in README.rst
- move files in models and views directories
- change headers to small license header
- use api 8.0
- fix and improve tests

yvaucher committed:
Update translations

yvaucher committed:
Cannot update options due to limitation of api 8.0

mdietrichc2c committed:
Use IDs instead of recordset for write

Commit:
rename field tracking_id to package_id

the field has been renamed in the base module

mdietrichc2c committed:
Remove self.pool + add correct attachment name

mdietrichc2c committed:
View changed in batch_stock_picking

Commit:
Fix condition

Commit:
Thread label generation to make it faster

Port of https://github.com/OCA/delivery-carrier/pull/27

mdietrichc2c committed:
Fix onchange for batch picking

mdietrichc2c committed:
Correct error treatment

Commit:
Fix empty pdf issue

Probably due to usage of multiple environment and cache. datas can't be
read from shipping.label but it can be read from ir.attachment

mdietrichc2c committed:
Add pick/pack info to all exceptions

mdietrichc2c committed:
e.message deprecated from 2.6

Commit:
Allow to print labels on operations without product

Commit:
Fix concurrent error on batch labels

When there is more than 1 pack in a picking, 2 or more thread worker
might be in concurrence on the same picking, which makes one of the
transaction fails, resulting in the abort of the whole label generation.

The solution implemented is to change the way of passing the tasks to
the thread workers. Instead of passing a label at a time, we pass a
group of label. A group of label actually contains all the labels to
generate for one picking.

Commit:
Do not regenerate existing labels unless forced + pylint fix

TDu committed:
[12.0][MIG] delivery_carrier_label_batch

rousseldenis commented:
I think the trend is to use queue jobs :

https://github.com/OCA/queue/tree/12.0/queue_job

@yvaucher @guewen What do you think as Camptocamp  was the original writer of this module ?

Commit:
[9.0][ADD] support of zpl labels

sebalix committed:
[IMP] delivery_carrier_label_batch: add support for ZPL labels

rousseldenis commented:
@TDu

tschanzt committed:
fixes for cahnges in migration of base_deliver_carrier_label (#1)

tschanzt commented:
@rousseldenis I checked and the problem is he uses threads to generate each individual label, but he needs them to return the values for the merge. If we use queue job we don't have control to do this. There is a project by @guewen to add dependencies in queue jobs but it is stalled.

sebalix commented:
@rousseldenis @tschanzt I talked to @guewen and the future job dependencies API won't support a `join/await` operation (https://github.com/OCA/delivery-carrier/pull/222/files#diff-b755b9f4f2e7cdfb81fa56517273d8ecR161) of jobs from the main process (one difficulty is to get job records committed in the middle of a transaction to get them running...).
So we have to keep `threading` here for this use case.

sebalix commented:
@TDu some developments have been done on 9.0 since this PR has been created (check https://github.com/OCA/delivery-carrier/pull/99/commits)
The following commits are missing:

- aba3b156c9880d7509dd24b8943bf16802e56545
- c9ea6329284333bd86ada33c9419d9b3b8d598b5

EDIT: the PR https://github.com/OCA/stock-logistics-workflow/pull/570 has been merged

rousseldenis commented:
> @rousseldenis I checked and the problem is he uses threads to generate each individual label, but he needs them to return the values for the merge. If we use queue job we don't have control to do this. There is a project by @guewen to add dependencies in queue jobs but it is stalled.

Ok, I see.

The point that makes me questionning is why do we need to manage threads ? What is the bottleneck in generation ?

sebalix commented:
@rousseldenis we need threads to generate labels in parallel and not sequentially to speed up the process when batch owns a lot of pickings.

sebalix committed:
[12.0][MIG] delivery_carrier_label_batch (fixes)

sebalix commented on delivery_carrier_label_batch/wizard/generate_labels.py:
```diff
@@ -86,9 +86,7 @@ def _do_generate_labels(self, group):
-                    picking.with_env(new_env).generate_labels(
-                        package_ids=package_ids
-                    )
+                    picking.with_env(new_env).action_generate_carrier_label()
```
@tschanzt This is a performance bottleneck here, before we were generating the label for a given package (`package_ids` above) but now we are generating the labels for all the picking packages in a `for` loop.
I do not know all the code, it needs to be rethought probably.
<s>If you want I can work on it.</s>

tschanzt replied:
@sebalix we stil have the threads per picking, so as we discussed in most cases the amount of packages per picking shouldn't be that high and doesn't require further measures.

Commit:
Fix tests

sebalix commented on delivery_carrier_label_batch/wizard/generate_labels.py:
```diff
@@ -0,0 +1,279 @@
+
+class DeliveryCarrierLabelGenerate(models.TransientModel):
+
+    _name = 'delivery.carrier.label.generate'
```
@tschanzt add `_description = "Generate labels from batch pickings"` to make runbot happy :)

Commit:
add missing _description

i-vyshnevska commented on delivery_carrier_label_batch/wizard/generate_labels.py:
```diff
@@ -0,0 +1,280 @@
+                picking = operations[0].picking_id
+                if pack:
+                    label = self_env._find_pack_label(pack)
+                else:
```
```suggestion
                elif not label and picking:
```
@sebalix should this condition check for both cases?

yvaucher replied:
IIRC we wants to enforce the use of packs and not try to print the labels if the packs are missings.

Testing locally having pickings without packs it wasn't working.

i-vyshnevska commented on delivery_carrier_label_batch/wizard/generate_labels.py:
```diff
@@ -0,0 +1,280 @@
+                if pack:
+                    label = self_env._find_pack_label(pack)
+                else:
+                    label = self_env._find_picking_label(picking)
```
```suggestion
                    picking = operations[0].picking_id
                    label = self_env._find_picking_label(picking)
```

yvaucher replied:
to remove

i-vyshnevska commented on delivery_carrier_label_batch/wizard/generate_labels.py:
```diff
@@ -0,0 +1,280 @@
+            self_env = self.with_env(new_env)
+            labels = []
+            for pack, operations, label in self_env._get_packs(batch):
+                picking = operations[0].picking_id
```
```suggestion
```

TDu replied:
Actually the whole for loop could be summarized as 
```
for pack, operations, label in self_env._get_packs(batch):
    picking = operations[0].picking_id
    if picking and not label:
        label = self_env._find_picking_label(picking)
    if not label:                                                                                                                                                                                                                                                                                            
        continue
     labels.append((label.file_type, label.attachment_id.datas))
```
Because the `_find_pack_label` is already called in `_get_packs`
As it is only a migration...

grindtildeath committed:
[IMP] delivery_carrier_label_batch: ZPL2 single images

grindtildeath committed:
Apply suggestions from code review

Co-Authored-By: Alexandre Fayolle <alexandre.fayolle@camptocamp.com>

grindtildeath committed:
Fixes after reviews

vrenaville committed:
[IMP][ZPL] option to not merge Zpl file (printer failed to print to big files)

vrenaville committed:
[IMP] label name from batch tracking zpl

yvaucher committed:
Add a check that all picking have packs before generating them.

yvaucher commented on delivery_carrier_label_batch/wizard/generate_labels.py:
```diff
@@ -0,0 +1,280 @@
+            yield pack, list(grp_operations), pack_label
+
+    @api.model
+    def _find_picking_label(self, picking):
```
Following the RFC here https://github.com/OCA/delivery-carrier/issues/192

And the commit
https://github.com/OCA/delivery-carrier/pull/212/commits/ffaab9796f157274c67ce78ee735e43a5128f36e

Pack must be enforced, thus the method `_find_picking_label` can be dropped.

yvaucher commented on delivery_carrier_label_batch/wizard/generate_labels.py:
```diff
@@ -0,0 +1,280 @@
+                    picking.with_env(new_env).action_generate_carrier_label()
+                except Exception as e:
+                    # add information on picking and pack in the exception
+                    picking_name = _('Picking: %s') % picking.name
+                    pack_num = _('Pack: %s') % pack.name if pack else ''
```
This most likely should be a single translated string. Depending on the language the result might not be correctly translatable. (Right-to-left languages)

An other option would be to enforce the use or line return instead of having it one line.

yvaucher commented on delivery_carrier_label_batch/wizard/generate_labels.py:
```diff
@@ -0,0 +1,280 @@
+                    pack_num = _('Pack: %s') % pack.name if pack else ''
+                    # pylint: disable=translation-required
+                    raise exceptions.UserError(
+                        ('%s %s - %s') % (picking_name, pack_num, e)
```
```suggestion
                        ('%s\n%s\n\n%s') % (picking_name, pack_num, e)
```

yvaucher reviewed (changes requested):
I'll propose a sub PR for the removal of find_picking_label.

The PR can be found here: https://github.com/TDu/delivery-carrier/pull/6


EDIT: added PR

yvaucher committed:
label batch: Enforce use of packages

The support of picking without packages has been dropped following the RFC #192

Since ffaab97 commit `base_delivery_carrier_label` force the usage of
packages.

yvaucher commented on delivery_carrier_label_batch/wizard/generate_labels.py:
```diff
@@ -0,0 +1,314 @@
+        return {
+            'type': 'ir.actions.act_window_close',
+        }
+
+
```
One line to remove

```suggestion

```

yvaucher commented on delivery_carrier_label_batch/wizard/generate_labels.py:
```diff
@@ -0,0 +1,314 @@
+                        self.env['ir.attachment'].create(data)
+                else:
+                    labels_bin = [
+                        codecs.decode(label['data'], "base64") for label in labels if label
```
```suggestion
                        codecs.decode(label['data'], "base64")
                        for label in labels if label
```

yvaucher committed:
Fix rst format, not an inline emphasis

yvaucher commented:
/ocabot merge nobump

yvaucher commented:
tests of `delivery_auto_refresh` are broken, I don't see the reason here.

Investigating...

yvaucher commented:
unit tests of auto refresh fixed in #299

yvaucher commented:
/ocabot merge nobump
